### PR TITLE
Refactoring API_PATH to allow multiple version URLs

### PR DESCRIPTION
**Pull request by rshbhgrg**
refs #9866

* remove hardcoded `API_PATH` with method to calculate path based on version
* update scheduler to use the stable(v0.1) endpoint
* serve public file uses the stable(v0.1) endpoint

### CODE REVIEW

File: core/server/services/url/utils.js [hunk 1]
```diff
@@ -238,7 +249,7 @@ function urlFor(context, data, absolute) {
         // this will become really big
         knownPaths = {
             home: '/',
-            api: API_PATH,
+            api: getApiPath('stable'),
```
Comment (kirrg001) [inline core/server/services/url/utils.js]: Is knownPath with `api` used? 🤔

Comment (rishabhgrg) [reply core/server/services/url/utils.js]: Good point, there is a general `knownPaths[context]`, but it seems anyways when `context = 'api'` code will never reach that point. Not sure why its there in first place, I'll remove it from knownPaths if there was no particular historic reason as it seems its not useful.

File: core/server/config/overrides.json [hunk 2]
```diff
@@ -64,5 +64,12 @@
     },
     "maintenance": {
         "enabled": false
+    },
+    "api": {
+        "versions": {
+            "active": "v2",
```
Comment (kirrg001) [inline core/server/config/overrides.json]: How does it look when i ask for the active admin url?

Comment (rishabhgrg) [reply core/server/config/overrides.json]: Updated the structure to handle admin/client URLs

**Comment by kirrg001** [comment]
There is one more usage in server/api/index.js.

File: core/server/services/url/utils.js [hunk 3]
```diff
@@ -6,10 +6,21 @@ const moment = require('moment-timezone'),
...
+    BASE_API_PATH = '/ghost/api/',
     STATIC_IMAGE_URL_PREFIX = 'content/images';
 
+/**
+ * Returns API path combining base path and path for specific version asked or stable by default
+ * @param {string} version version for which to get the path(stable, actice, deprecated: client, admin), defaults to stable:client
+ * @return {string} API Path for version
+ */
+function getApiPath(version = 'stable', admin = false) {
+    const apiVersions = config.get('api:versions') || {};
```
[the remark this answers is not in the document]
Comment (rishabhgrg) [reply core/server/services/url/utils.js]: 👍 Makes sense

**Comment by rishabhgrg** [comment]
👍 Had left it to implicitly pick up stable version(v0.1) before as was not sure about location headers then, updated.

File: core/server/config/overrides.json [hunk 4]
```diff
@@ -64,5 +64,18 @@
...
+    },
+    "api": {
+        "versions": {
+            "active": {
+                "admin": "v2/admin",
+                "client": "v2/client"
+            },
+            "stable": {
+                "admin": "v0.1",
+                "client": "v0.1"
```
Comment (rishabhgrg) [inline core/server/config/overrides.json]: @kirrg001 Should we add another option for each API version as `"default"` other than `"admin"` and `"client"` for future-proofing? Currently either one needs to pass explicit `admin: true` or it defaults to client. Can be also something which we can add later if needed though, wdyt?

File: core/server/services/url/utils.js [hunk 5]
```diff
@@ -6,10 +6,21 @@ const moment = require('moment-timezone'),
...
 
+/**
+ * Returns API path combining base path and path for specific version asked or stable by default
+ * @param {string} version version for which to get the path(stable, actice, deprecated: client, admin), defaults to stable:client
+ * @return {string} API Path for version
+ */
+function getApiPath(version = 'stable', admin = false) {
+    const apiVersions = config.get('api:versions');
+    let versionType = apiVersions[version] || apiVersions.stable;
+    let versionPath = admin ? versionType.admin : versionType.client;
```
Comment (rishabhgrg) [inline core/server/services/url/utils.js]: Picks `stable:client` to be default version if nothing is passed

File: core/server/services/url/utils.js [hunk 6]
```diff
@@ -303,7 +313,7 @@ function urlFor(context, data, absolute) {
         }
     } else if (context === 'api') {
         urlPath = getAdminUrl() || getBlogUrl();
-
+        let apiPath = getApiPath('stable');
```
Comment (rishabhgrg) [inline core/server/services/url/utils.js]: Again assumes `stable` to be default API if no data is passed to `urlFor` with context.

File: core/test/unit/services/url/utils_spec.js [hunk 7]
```diff
@@ -522,6 +522,22 @@ describe('Url', function () {
...
+            urlService.utils.urlFor('api', {cors: true, version: "active"}, true).should.eql('https://my-ghost-blog.com/ghost/api/v2/client/');
+        });
+
+        it('api: with active version and admin true, blog url is https: should return active admin api path', function () {
+            configUtils.set({
+                url: 'https://my-ghost-blog.com'
+            });
+
+            urlService.utils.urlFor('api', {cors: true, version: "active", admin: true}, true).should.eql('https://my-ghost-blog.com/ghost/api/v2/admin/');
+        });
```
Comment (rishabhgrg) [inline core/test/unit/services/url/utils_spec.js]: Didn't add any explicit test to check for URL returned for `version: "stable"` as it gets tested with other tests implicitly right now. Can be updated with tests upgrade task

File: core/server/config/overrides.json (hunk 4 above)
Comment (kirrg001) [reply core/server/config/overrides.json]: Hmm i think we can add this if we really need it?

File: core/server/config/overrides.json [hunk 8]
```diff
@@ -64,5 +64,18 @@
     },
     "maintenance": {
         "enabled": false
+    },
+    "api": {
+        "versions": {
+            "active": {
+                "admin": "v2/admin",
+                "client": "v2/client"
```
Comment (kirrg001) [inline core/server/config/overrides.json]: `/client`? 🤔

File: core/server/services/url/utils.js (hunk 6 above)
Comment (kirrg001) [reply core/server/services/url/utils.js]: I think that is fine, even though i like that you currently explicitly pass `stable:true` 👍

File: core/server/config/overrides.json (hunk 8 above)
Comment (rishabhgrg) [reply core/server/config/overrides.json]: Haha! 🤐 Not sure what I was thinking about `client` while editing this, for some reason that just stuck out of brain even here! Updating this silly 🐛

File: core/server/config/overrides.json (hunk 4 above)
Comment (rishabhgrg) [reply core/server/config/overrides.json]: Yeah was thinking of same, no need for premature optimization as for near future we only have `content` and `admin`!

File: core/test/unit/services/url/utils_spec.js (hunk 7 above)
Comment (kirrg001) [reply core/test/unit/services/url/utils_spec.js]: Maybe add one example with `stable:true`?

Comment (rishabhgrg) [reply core/test/unit/services/url/utils_spec.js]: Sure :)

**Comment by kirrg001** [comment]
PR looks good to me, will test & merge later 👍

**Review by kirrg001 — APPROVED** [review]
👍

**Pull request merged**